### PR TITLE
[Hardware][Gaudi][Feature] Enable Dynamic MoE for Mixtral

### PR DESCRIPTION
**Pull request by zhenwei-intel**
## Enable Dynamic MoE for Mixtral

Command for testing accuracy:
`bash .buildkite/lm-eval-harness/run-lm-eval-gsm-hf-baseline.sh -m neuralmagic/Mixtral-8x7B-Instruct-v0.1 -b 32 -l 250 -f 5 -t 4`

Accuracy results for bf16 precision:
image

### CODE REVIEW

**Comment by zhenwei-intel** [comment]
@kzawora-intel @mgoin, Hi, could you help review this pr?

File: vllm/model_executor/models/mixtral.py [hunk 1]
```diff
@@ -501,4 +502,6 @@ def load_weights(self, weights: Iterable[Tuple[str,
                                             default_weight_loader)
                     weight_loader(param, loaded_weight)
             loaded_params.add(name)
+            if current_platform.is_hpu():
+                torch.hpu.synchronize()
```
Comment (mgoin) [inline vllm/model_executor/models/mixtral.py:506]: Is this something you specifically need for MoEs? I would like you to avoid touching model definitions, since we would have to do this to every MoE model. Hopefully you can move this into your hpu code

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 2]
```diff
@@ -186,6 +186,26 @@ def forward_cpu(
...
+        num_expert_group: Optional[int] = None,
+        **kwargs,
+    ):
+        assert not use_grouped_topk, "use_grouped_topk must be False on HPU"
+        assert num_expert_group is None, ('num_expert_group is '
+                                          'not supported on HPU')
+        assert topk_group is None, "topk_group is not supported on HPU"
+        if layer is not None:
+            return layer.hpu_fused_moe(x, layer.w13_weight, layer.w2_weight,
+                                       router_logits, top_k)
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/layer.py:207]: Please assert/check the other kwargs i.e.
```
        custom_routing_function: Optional[Callable] = None,
        scoring_func: str = "softmax",
        e_score_correction_bias: Optional[torch.Tensor] = None
```

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 3]
```diff
@@ -186,6 +186,26 @@ def forward_cpu(
...
+        renormalize: bool,
+        topk_group: Optional[int] = None,
+        num_expert_group: Optional[int] = None,
+        **kwargs,
+    ):
+        assert not use_grouped_topk, "use_grouped_topk must be False on HPU"
+        assert num_expert_group is None, ('num_expert_group is '
+                                          'not supported on HPU')
+        assert topk_group is None, "topk_group is not supported on HPU"
+        if layer is not None:
```
Comment (mgoin) [inline vllm/model_executor/layers/fused_moe/layer.py:205]: Why isn't this an assert?

File: vllm/model_executor/models/mixtral.py (hunk 1 above)
Comment (zhenwei-intel) [reply vllm/model_executor/models/mixtral.py]: Thanks, fix oom issue in another way: 98a07f31c0929e6d84e64c1036cc8b8a57c40b67

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 2 above)
Comment (zhenwei-intel) [reply vllm/model_executor/layers/fused_moe/layer.py]: updated this function bfe5d0cba0926cbfc1461e9edb187c02649321c2

**Comment by zhenwei-intel** [comment]
Hi @mgoin, could you review the updated code again?

File: vllm/utils.py [hunk 4]
```diff
@@ -353,6 +353,11 @@ def reset(self):
         self._index = 0
 
 
+@cache
+def is_fake_hpu() -> bool:
+    return os.environ.get('VLLM_USE_FAKE_HPU', '0') != '0'
```
Comment (WoosukKwon) [inline vllm/utils.py:358]: Could you move it to `hpu_utils.py`?

Comment (zhenwei-intel) [reply vllm/utils.py]: thanks, removed it.

**Comment by zhenwei-intel** [comment]
Hi @mgoin @WoosukKwon , could you help review again?

**Review by WoosukKwon — APPROVED** [review]

**Review by WoosukKwon — APPROVED** [review]
LGTM. Sorry for the delay.
